busybox: add getedid script for RPi

HiassofT opened the pull request:
This script simply creates /flash/edid.cpio with the contents from
.config/firmware/edid

Compared to the x86 specific getedid script this works on all platforms
as it doesn't contain any platform specific code. Eventually getedid
should be refactored make use of this helper script, too.

lrusak commented on packages/sysutils/busybox/scripts/create-edid-cpio:
```diff
@@ -0,0 +1,28 @@
+# SPDX-License-Identifier: GPL-2.0
+# Copyright (C) 2021-present Team LibreELEC (https://libreelec.tv)
+
+EDID_DIR="/storage/.config/firmware/edid"
```
I don't really understand, so you need to have an edid already to make an edid cpio?

Why can't you get the edid from the drm subsystem via a valid connector?

We already have the instructions here -> https://libreelec.wiki/configuration/edid

HiassofT commented:
The purpose of the script is only to make /usr/lib//firmware/edid to initramfs so it's available when the drm driver probes - to get something on screen early during boot.

Storing the appropriate edid(s) in .config/firmware/edid is the responsibility of the user ATM - this may be later done by another script.

Modifying bootloader to load the cpio as initramfs is yet another independent thing - and as this highly dependent on bootloader (grub/syslinux/uboot/rpi) that should be done by yet another script. Same goes for modifying cmdline.txt to append the appropriate drm.edid_firmware and video parameters.

The current getedid script is too closely tied to x86 and also doesn't handle multiple HDMI connectors - users on RPi4 want to make use of both HDMI ports, one for TV, another one for AVR and both would need edids and video settings, maybe also overriding the actual video modes.

As this is tricky and we don't yet know all the possible things users might want to do I started with one part that is common to all platforms, corrects the issue in getedid and wiki (putting the edid in /lib/firmware instead of /usr/lib/firmware results in non-booting 32bit systems as that trashes the /lib->/usr/lib symlink) and can simply be called from a future getedid replacement.

lrusak commented:
> The purpose of the script is only to make /usr/lib//firmware/edid to initramfs so it's available when the drm driver probes - to get something on screen early during boot.
> 
> 
> 
> Storing the appropriate edid(s) in .config/firmware/edid is the responsibility of the user ATM - this may be later done by another script.
> 
> 
> 
> Modifying bootloader to load the cpio as initramfs is yet another independent thing - and as this highly dependent on bootloader (grub/syslinux/uboot/rpi) that should be done by yet another script. Same goes for modifying cmdline.txt to append the appropriate drm.edid_firmware and video parameters.
> 
> 
> 
> The current getedid script is too closely tied to x86 and also doesn't handle multiple HDMI connectors - users on RPi4 want to make use of both HDMI ports, one for TV, another one for AVR and both would need edids and video settings, maybe also overriding the actual video modes.
> 

I disagree, the current getedid script already does everything we need other than modifying the boot loader. I would suggest adding the functionality to get the edid for any drm driver to the getedid script, then it can be made to bail out in the case of uboot/RPi and that functionality can be added later.

> 
> As this is tricky and we don't yet know all the possible things users might want to do I started with one part that is common to all platforms, corrects the issue in getedid and wiki (putting the edid in /lib/firmware instead of /usr/lib/firmware results in non-booting 32bit systems as that trashes the /lib->/usr/lib symlink) and can simply be called from a future getedid replacement.

If the wiki is wrong please fix it.

I really don't think we need another method of fetching an edid (especially when this one really doesn't do anything).

Please put the time in to fix the script we already have as it will benefit all users.

HiassofT commented:
> I disagree, the current getedid script already does everything we need other than modifying the boot loader. I would suggest adding the functionality to get the edid for any drm driver to the getedid script, then it can be made to bail out in the case of uboot/RPi and that functionality can be added later.

I strongly disagree. Please read through the getedid script and what I wrote. getedid doesn't handle the case where we have 2 edids for 2 connectors - getedid only handles a single edid.bin - which won't help for the RPi4 with 2 active hdmi connectors, each needing it's own edid.

As I can't test x86/AMD/Nvidia I'm also quite reluctant to touch the x86-only getedid script or change the x86 specific wiki page - I pointed out the issues to davu on slack several months ago.

And as more RPi users are popping up on the forum with edid issues we need to start somewhere. Following the unix philosophy "done one thing, and to it good" seems like a very good start to me.

lrusak commented:
> > I disagree, the current getedid script already does everything we need other than modifying the boot loader. I would suggest adding the functionality to get the edid for any drm driver to the getedid script, then it can be made to bail out in the case of uboot/RPi and that functionality can be added later.
> 
> 
> 
> I strongly disagree. Please read through the getedid script and what I wrote. getedid doesn't handle the case where we have 2 edids for 2 connectors - getedid only handles a single edid.bin - which won't help for the RPi4 with 2 active hdmi connectors, each needing it's own edid.
> 

This is a really dumb use case to even try to provide a solution for but I understand that people may want to do this for their setups.

People may also want to do that with their x86 machine.

The drm subsystem is identical for everything but nvidia so making any changes to the getedid script should be trivial.

We really don't need an edid script for each platform (I really don't think we need one at all). It really should just be documented in the wiki with the correct procedure rather than having half working scripts all over the place.

> 
> As I can't test x86/AMD/Nvidia I'm also quite reluctant to touch the x86-only getedid script or change the x86 specific wiki page - I pointed out the issues to davu on slack several months ago.
> 
> 
> 
> And as more RPi users are popping up on the forum with edid issues we need to start somewhere. Following the unix philosophy "done one thing, and to it good" seems like a very good start to me.

Seems like a workaround more than anything. Maybe make a forum sticky with what users should do?

HiassofT commented:
getedid can simply be refactored to make use of this new script - it's not intended to be a getedid replacement on it's own but help users with the most involved part (most commands to enter) of getting edid override working.

Refactoring getedid to cover all the usecases we currently have isn't exactly trivial. And instead of possibly breaking getedid it's a lot safer to factor out common stuff and develop that in parallel.

While intel/amd DRM is identical to others, nvidia isn't (and it needs tons of code for xorg handling in getedid, too). Same goes for bootloader stuff - so that is better handled in separate scripts so we don't end up with an unmaintainable single monster script.

HiassofT commented:
I've now implemented the missing helper scripts to dump edids of active connectors and modify RPi's bootloader config and added a "getedid" script that works identical to the x86 one:

"getedid create" dunps edids and adds edid overrides in bootloader, "getedid delete" removes bootloader override. I choose to keep dumped firmwares and edid.cpio in place though on delete.

The "getedid" script should work on other (DRM) platforms as well as it doesn't contain any platform specific code. Once u-boot config has been implemented for AW/RK/... via a separate script getedid can be installed on those platforms as well.

lrusak commented:
Thanks for taking the time to actually implement the functionality. :+1: 

I do like the abstraction and the consideration for other platforms. :1st_place_medal: 

I'm a bit sceptical if this is the best approach though. As you now provide 4 scripts your provide 4 entry points, which is the correct one to call? It's a little bit confusing to call out to other scripts from within a script that also has functions. Why not source the other scripts from the main script to create only one entry point? There is also no safety in case of the other scripts being missing (which may or may not happen).

HiassofT commented:
Entry point for users is only getedid - and it checks the return code of the helper scripts so if they are missing they'll get an error message.

One reason to keep things separate is that I often just need create-edid-cpio (which I initially PR'ed) - I regularly test with different edids, add new ones to .config/firmware/edid and manually adjust kernel cmdline.

Adding a couple of options to getedid would be another possibility, but having separate scripts that can be invoked individually is a bit safer as then normal users won't find out about these easily and are less likely to shoot in their feet.
